ContosoPizza/Features/Users/Add/AddUserHandler.cs

```diff
@@ -0,0 +1,53 @@
+﻿using ContosoPizza.Models;
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+using Nudes.Retornator.AspnetCore.Errors;
+using Nudes.Retornator.Core;
+using System.Security.Claims;
+
+namespace ContosoPizza.Features.Users.Add
+{
+    public class AddUserHandler : IRequestHandler<AddUserRequest, ResultOf<int>>
+    {
+        private readonly ContosoPizzaContext db;
+        private readonly IHttpContextAccessor httpContextAccessor;
+
+        public AddUserHandler(ContosoPizzaContext db, IHttpContextAccessor httpContextAccessor)
+        {
+            this.db = db;
+            this.httpContextAccessor = httpContextAccessor;
+        }
+
+        public async Task<ResultOf<int>> Handle(AddUserRequest request, CancellationToken cancellationToken)
+        {
+            var UserContext = httpContextAccessor.HttpContext.User;
+            var roleId = UserContext.Claims.FirstOrDefault(claim => claim.Type == ClaimTypes.Role);
+
+            if (!await db.Roles.AnyAsync(role => role.Id == request.RoleId, cancellationToken)) //se o role da request nao existe
+            {
+                return new BadRequestError();
+            }
+            var idCliente = (await db.Roles.FirstOrDefaultAsync(d => d.Name == "cliente", cancellationToken)).Id;
+
+            if (roleId == null || roleId.Value == idCliente.ToString()) //se o usuaro for anonimo ou um cliente
```

Comment:
se role Id for nulo nós carregamos já o id do role "cliente" isso é performático apesar de funcionar.
Basear-se com o Role em texto hardcoded também é um ponto ruim

porém o pior mesmo é que se adicionarmos um novo role que não deveria ter essa permissão e esse role faz o cadastro nós passariamos. A verificação deveria ser `se o usuário for um funcionário` e não `se o usuário não for um cliente`